x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/planner/Verifier.java

```diff
@@ -51,4 +57,24 @@
 
         return failures;
     }
+
+    private static void checkForNonCollapsableSubselects(PhysicalPlan plan, List<Failure> failures) {
+        Holder<Boolean> hasLimit = new Holder<>(Boolean.FALSE);
+        Holder<List<Order>> orderBy = new Holder<>();
+        plan.forEachUp(p -> {
+            if (hasLimit.get() == false && p instanceof LimitExec) {
+                hasLimit.set(Boolean.TRUE);
+                return;
+            }
+            if (orderBy.get() == null && p instanceof OrderExec) {
+                orderBy.set(((OrderExec) p).order());
+                return;
+            }
+            if (hasLimit.get() && orderBy.get() != null && p instanceof OrderExec) {
+                if (((OrderExec) p).order().equals(orderBy.get()) == false) {
+                    failures.add(fail(p, "Cannot use ORDER BY on top of a subquery with ORDER BY and LIMIT"));
+                }
+            }
```

Comment:
Optional compacting.

```suggestion
            if (p instanceof LimitExec) {
                if (hasLimit.get() == false) {
                    hasLimit.set(true); // I'd use either objects or primitives for this and above statement
                }
            } else if (p instanceof OrderExec) {
                if (orderBy.get() == null) {
                    orderBy.set(((OrderExec) p).order());
                } else if (hasLimit.get()) {
                    if (((OrderExec) p).order().equals(orderBy.get()) == false) {
                        failures.add(fail(p, "Cannot use ORDER BY on top of a subquery with ORDER BY and LIMIT"));
                    }
                }
            }
```